src/Model_Hydro/CPU_Hydro/CPU_Shared_AddSourceTerm.cpp

```diff
@@ -0,0 +1,458 @@
+//    const real B_xR = g_FC_B_In[0][idx_B_xR];
+//    const real B_yL = g_FC_B_In[1][idx_B_yL];
+//    const real B_yR = g_FC_B_In[1][idx_B_yR];
+//    const real B_zL = g_FC_B_In[2][idx_B_zL];
+//    const real B_zR = g_FC_B_In[2][idx_B_zR];
+#  endif
+
+// Example: cosmic ray adiabatic work done term
+// #  ifdef COSMIC_RAY
+// // 1. calculate the cosmic-ray pressure
+//    const real pCR_old = EoS->CREint2CRPres_FuncPtr( g_ConVar_In[CRAY][idx_in], EoS->AuxArrayDevPtr_Flt,
+//                                                     EoS->AuxArrayDevPtr_Int, EoS->Table );
+//
+//
+// // 2. compute \div V using the upwind data; reference: [2]
+//    real div_V[3];
+//
+//    for (int d=0; d<3; d++)
+//    {
+//       div_V[d] = (real)0.5 * ( g_ConVar_In[DENS+d][idx_in + didx_in[d]] / g_ConVar_In[DENS][idx_in + didx_in[d]] -
+//                                g_ConVar_In[DENS+d][idx_in - didx_in[d]] / g_ConVar_In[DENS][idx_in - didx_in[d]] );
+//    } // for (int d=0; d<3; d++)
+//
+//
+// // 3. update the cosmic-ray energy
+//    OneCell[CRAY] -= pCR_old*dt_dh2*( div_V[0] + div_V[1] + div_V[2] );
+// #  endif // #ifdef COSMIC_RAY
+
+} // FUNCTION : Hydro_AddSourceTerm_CCVar_HalfStep_MHM_RP
+
+
+
+#ifdef MHD
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Hydro_AddSourceTerm_FCVar_HalfStep_MHM_RP
+//
+// Description :  Add source term on the face-centered magnetic field at half-step update
+//
+// Note        :  1. This function can only be used by MHM_RP
+//                2. Invoked by CPU/CUFLU_FluidSolver_MHM()
+//                3. Must be used after MHD_UpdateMagnetic()
+//
+// Parameter   :  g_ConVar_In : Array storing the input conserved variables
+//                g_FC_B_In   : Array storing the input     face-centered magnetic field (for MHD only)
+//                g_FC_B_Half : Array storing the half-step face-centered magnetic field (for MHD only)
+//                dt          : Time interval to advance solution
+//                dh          : Cell size
+//                EoS         : EoS object
+//
+// Return      :  g_FC_B_Half
+//-------------------------------------------------------------------------------------------------------
+GPU_DEVICE
+void Hydro_AddSourceTerm_FCVar_HalfStep_MHM_RP( const real g_ConVar_In[][ CUBE(FLU_NXT) ],
+                                                const real g_FC_B_In[][ FLU_NXT_P1*SQR(FLU_NXT) ],
+                                                      real g_FC_B_Half[][ FLU_NXT_P1*SQR(FLU_NXT) ],
+                                                const real dt, const real dh, const EoS_t *EoS )
+{
+
+#  ifdef GAMER_DEBUG
+   if ( FLU_NXT-N_HF_VAR != 2 )   printf( "FLU_NXT(%d) - N_HF_VAR(%d) != 2 !\n", FLU_NXT, N_HF_VAR );
+#  endif
+
+   const real dt_dh2      = (real)0.5 * dt / dh;
+   const int  N_HF_VAR_P1 = N_HF_VAR + 1;
+   const int  didx_in[3]  = { 1, FLU_NXT, SQR(FLU_NXT) };
```

Comment:
How about renaming this variable to better reflect the fact that it's for accessing `g_ConVar_In[]`?